CP-33774 PSR orchestration

lippirk opened the pull request:
We implement a function which orchestrates a pool secret rotation.
We handle failure by splitting the PSR into idempotent chunks, and
saving checkpoints at the end of each chunk. This will allow us to
resume from failed PSRs, and importantly reduce the chance of hosts
becoming locked out with the wrong pool secret.

Signed-off-by: lippirk <ben.anson@citrix.com>

lindig commented on ocaml/tests/suite_alcotest.ml:
```diff
@@ -75,4 +75,5 @@ let () =
       @ Test_helpers.tests
-      @ Test_datamodel_utils.tests);
+      @ Test_datamodel_utils.tests
+      @ Test_psr.tests );
```
It doesn't matter much here, but `List.concat` would be more efficient. We also had some nasty surprised when using `@` together with if-then-else.

lindig commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,122 @@
+  type ps = string
+  type pss = (ps * ps)
+  type fistpoint = No_fp
+                 | Accept_new_ps
```
The module uses an acronym (psr) and `ps` is used here, too. Looking from the outside it might be good to spell out `pool_secret` somewhere.

lippirk replied:
i mentioned the abbreviations in the mli file, but you are right - this is an indication i should change the name of the type

lindig commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,142 @@
+  let retrieve () = Option.get !backups
+
+  let tell_accept_new_ps (_, staged_ps) h =
+    if !h.fistpoint = Accept_new_ps then
```
How about
```
match !h.fistpoint with
| Accept_new_ps ->
| .. ->
```

I think this is easier to read and works better with algebraic types.

lindig commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,142 @@
+
+let test_fails_twice () =
+  let hosts = Impl.mk_hosts 3 in
+  let faulty_host_1 = List.nth hosts 2 in
```
Slightly confusing to select host 2 but call it host 1.

lindig commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,142 @@
+
+let backups = ref None
+
+let default_ps = "default_ps"
```
The code uses acronyms in identifiers for modules, types, and values: ps and cp. Looking from the outside, this can be hard to read. I think some more explicit names for modules would help the reader.

psafont commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,159 @@
+
+  let backup pool_secrets = backups := Some pool_secrets
+
+  let retrieve () = Option.get !backups
```
This will raise an exception if it gets called before `let backup` is called, is it not better to return an option so the caller can deal with it?

lippirk replied:
i wrote `val retrieve : unit -> pool_secrets option` originally, but it actually doesn't make much sense. if we can't fetch the pool secret from the backup that's a fatal error (i.e. the implementor should handle that). the reason it's a fatal error is that we fetch from the backup only when there is a checkpoint to resume from (you'll see that the function to get the checkpoint returns an option)

there are racy, and i ran into this when writing the tests :-1: i think this is somewhat useful here because i do want to make sure that the pool secret backup is cleaned up (for example). i'm not sure what i'd use as an alternative to model writing/reading to a file?

edwintorok commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,143 @@
+
+    exception Cannot_parse_checkpoint of string
+
+    let string_of_checkpoint = function
```
You can use the RPC derivers to generate serialization and deserialization functions for you, avoiding all the boilerplate.

edwintorok commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,143 @@
+          | None ->
+              (go [@tailcall]) pool_secrets hosts Send_new_pool_secret
+          | Some e ->
+              save_checkpoint (string_of_checkpoint Accept_new_pool_secret) ;
```
I think the checkpoint should be saved when you enter the go function, prior to doing the matching, that way you don't need all these match None/Some on each state, and also you would save a checkpoint even if xapi or the host crashes in the middle of running the operation.

As for implementing the save checkpoint: perhaps the redolog code can be used here? That already has necessary logic for recovering from failures, etc. We could require the redolog to be enabled (which doens't necessarily mean that HA has to be enabled, in fact HA should be disabled...) before doing a PSR

lippirk commented:
unfortunately the tests have become more complicated because the slaves and the master are now distinguished (the reason for doing this was that i found in the real implementation it matters what order you cleanup the hosts after the rotation has finished). i also increased the coverage by testing more fistpoint combinations

lippirk commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,350 @@
+  | Failed_during_cleanup_master x ->
+      Printf.sprintf "Failed_during_cleanup_master: %s" (string_of_int x)
+
+let map_r f = function
```
i added an extra variant in my other PR (#4127) - `UnknownError of exn`, which makes it harder to model with result (but i agree in its form here to use `result` would definitely have been better)

lippirk commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,350 @@
+
+let string_of_fistpoint_time = function Before -> "Before" | After -> "After"
+
+type firstpoint_place =
```
i renamed this and added a comment to make it clearer

robhoes commented on ocaml/xapi/xapi_psr.mli:
```diff
@@ -0,0 +1,60 @@
+  | Failed_during_accept_new_pool_secret of 'a
+  | Failed_during_send_new_pool_secret of 'a
+  | Failed_during_cleanup_slave of 'a
+  (* very unlikely - user needs to cleanup manually *)
```
We can't tell users to delete files in dom0. We should just keep trying, for example on startup.

lippirk replied:
:heart: that will make the code simpler

robhoes commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,143 @@
+  let name = "xapi_psr"
+end)
+
+type 'a r =
```
I think that it is a good idea to make use of the `result` type for this kind of thing. This is now the standard. You could then also replace the use of exceptions below with a monadic style around `result` (see Rresult).

robhoes commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,143 @@
+  (Impl : Impl)
+  ->
+  struct
+    open Impl
```
I think that not opening will make the code clearer.

robhoes commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,143 @@
+          (go [@tailcall]) pool_secrets hosts Cleanup_slaves
+      | Cleanup_slaves ->
+          save_checkpoint (string_of_checkpoint Cleanup_slaves) ;
+          let[@warning "-8"] (master :: slaves) = hosts in
```
Perhaps have `master` and `slaves` as arguments to `go`, so you don't need this?

lippirk commented:
have addressed the comments, notable changes:

 - use result type
 - slave -> member
 - put master cleanup in a retry loop (this also changes the unit test a little)

robhoes commented on ocaml/xapi/xapi_psr.ml:
```diff
@@ -0,0 +1,157 @@
+          (go [@tailcall]) pool_secrets master members Cleanup_master
+      | Cleanup_master ->
+          Impl.save_checkpoint (string_of_checkpoint Cleanup_master) ;
+          Helpers.retry_forever ~interval:1. ~doc:"cleanup master after PSR"
```
This means that the API call will potentially hang forever, and the caller will not be able to cancel it. What is the consequence of this failure state? Everything works with the new token, but the master still accepts the old one? It may be better to treat it like the other failure cases: fail with and error describing the situation and let the user retry the call.

psafont commented on ocaml/tests/test_psr.ml:
```diff
@@ -0,0 +1,340 @@
+ *)
+let almost_all_possible_fists ~num_hosts =
+  let multiply fp_times fp_actions host_ids =
+    List.map
```
Maybe next time we'll be able to use `List.concat_map`...

lippirk replied:
:crossed_fingers:
